components/brave_search/browser/brave_search_default_host.cc

```diff
@@ -0,0 +1,141 @@
+// Copyright (c) 2021 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// you can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "brave/components/brave_search/browser/brave_search_default_host.h"
+
+#include <string>
+#include <utility>
+
+#include "brave/components/brave_search/browser/prefs.h"
+#include "brave/components/weekly_storage/daily_storage.h"
+#include "components/prefs/pref_service.h"
+#include "components/search_engines/template_url.h"
+#include "components/search_engines/template_url_service.h"
+
+namespace {
+
+TemplateURL* GetSearchTemplateForSite(TemplateURLService* service,
+                                      const std::string& host) {
+  // Prefer built-in entries, then offer site-provided entries,
+  // but ignore extension-provided entries.
+  TemplateURL::TemplateURLVector urls = service->GetTemplateURLs();
+  TemplateURL* other_entry = nullptr;
+  for (auto* template_url : urls) {
+    if (template_url->url_ref().GetHost(SearchTermsData()) != host) {
+      continue;
+    }
+    if (service->ShowInDefaultList(template_url)) {
+      return template_url;
+    } else if (template_url->type() != TemplateURL::OMNIBOX_API_EXTENSION &&
+               other_entry == nullptr) {
+      other_entry = template_url;
+    }
+  }
+  return other_entry;
+}
+
+}  // namespace
+
+namespace brave_search {
+
+BraveSearchDefaultHost::BraveSearchDefaultHost(
+    const std::string& host,
+    TemplateURLService* template_url_service,
+    PrefService* prefs)
+    : host_(host),
+      template_url_service_(template_url_service),
+      prefs_(prefs),
+      weak_factory_(this) {}
+
+BraveSearchDefaultHost::~BraveSearchDefaultHost() {}
+
+bool BraveSearchDefaultHost::CanSetDefaultSearchProvider(TemplateURL* provider,
+                                                         bool is_historic) {
+  // Return false if:
+  //   - current site is default OR
+  //   - has asked more than once today OR
+  //   - is incognito / private OR
+  // Otherwise return true
+  if (!provider) {
+    VLOG(1) << "CanSetDefaultSearchProvider: "
+            << "Site has not created search engine: " << host_;
+    return false;
+  }
+  const auto* default_provider =
```

Comment:
Recently, I pushed crash fix (https://github.com/brave/brave-core/pull/8930) to handle `nullptr` return value from `TemplateURLService::GetDefaultSearchProvider()`.
I'm not 100% sure that crash came from this `nullptr`. but I think it would be good to handle it here also.

Reply:
fixed